### PR TITLE
Purchase item

### PR DESCRIPTION
**Pull request by andarcabrera**
Here is the pull request for this story:
- I can click "Buy it now" on any item
- I can enter my email address to complete the transaction
- An email is sent to the seller's email address with the buyer's email address
- A third party is sending the email to the seller with the buyer's email in the body
- The item is removed from the collection of listings
- Mark item as unavailable

We also improved? our UI

Looking forward to your comments.

Anda & Jerome

### CODE REVIEW

File: app/assets/stylesheets/items.scss [hunk 1]
```diff
@@ -0,0 +1,59 @@
+.jumbotron {
+  background-color: #94618E !important;
```
Comment (kevbuchanan) [inline app/assets/stylesheets/items.scss]: Don't use important: https://css-tricks.com/when-using-important-is-the-right-choice/

Comment (kevbuchanan) [inline app/assets/stylesheets/items.scss]: If you need to use important, there's usually a better way to accomplish what you need.

File: config/routes.rb [hunk 2]
```diff
@@ -1,5 +1,7 @@
 Rails.application.routes.draw do
  root 'items#index'
- get  "/items/new" =>  "items#new"
- post "/items"     =>  "items#create"
+ get  "/items/new"                =>  "items#new"
+ post "/items"                    =>  "items#create"
+ get  "/items/:id/purchases/new"  =>  "purchases#new"
+ post  "/purchases"               =>  "purchases#create"
```
Comment (kevbuchanan) [inline config/routes.rb]: Spacing is off.
- spacing is fixed

File: config/routes.rb [hunk 3]
```diff
@@ -1,5 +1,7 @@
 Rails.application.routes.draw do
  root 'items#index'
- get  "/items/new" =>  "items#new"
- post "/items"     =>  "items#create"
+ get  "/items/new"                =>  "items#new"
+ post "/items"                    =>  "items#create"
+ get  "/items/:id/purchases/new"  =>  "purchases#new"
```
Comment (kevbuchanan) [inline config/routes.rb]: Why is "purchases#new" nested under items but "purchases#create" is not?

File: app/controllers/purchases_controller.rb [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    purchase = new_purchase(item_id: item_id)
+    @presenter = create_presenter(purchase, item)
+  end
+
+  def create
+    purchase = new_purchase(purchase_params)
+    item = find_item(purchase_params[:item_id])
+    @presenter = create_presenter(purchase, item)
+
+    if @presenter.purchase.save
```
Comment (kevbuchanan) [inline app/controllers/purchases_controller.rb]: Why do you go through the presenter to save the purchase? Why not `purchase.save` ?

File: app/models/item.rb [hunk 5]
```diff
@@ -1,5 +1,11 @@
 class Item < ActiveRecord::Base
   validates :name, :description, :price, :email,  presence: true
   mount_uploader :image, ImageUploader
+
+  before_create :make_available
```
Comment (kevbuchanan) [inline app/models/item.rb]: Why is this needed if the column default is true?
- That's what we thought initially, but we couldn't get it working. We made the change and it works now. maybe we just needed to restart the server?

File: app/controllers/purchases_controller.rb [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+      item.update_attributes(available: false)
```
Comment (kevbuchanan) [inline app/controllers/purchases_controller.rb]: You have a race condition in the way purchases are made and the available attribute is used. What if two people are looking at the index page, one makes a purchase, and then the other purchases the same item?
- Without adding an index to the purchases table, there is no guarantee of data integrity. If  two users see the same item as available, the purchase for both would go through. This would give us invalid data in our table and the users might pay the seller twice for an item that only one of them actually purchased. By adding an index we ensure that one of the transactions rolls back after hitting the uniqueness constraints now imposed on the data.

File: lib/item_marketplace/presenters/items_index_presenter.rb [hunk 7]
```diff
@@ -5,4 +5,8 @@ class ItemsIndexPresenter
   def initialize(items)
     @items = items
   end
+
+  def available_items
+    @items.select {|item| item.available == true}
```
Comment (kevbuchanan) [inline lib/item_marketplace/presenters/items_index_presenter.rb]: In ActiveRecord, boolean attributes give you the method "attribute?", so this could more concisely be written as `@items.select(&:available?)`

Comment (kevbuchanan) [inline lib/item_marketplace/presenters/items_index_presenter.rb]: Also, you'll probably want to pass in available items to the presenter. Unless the presenter will be querying based on any future pagination params. For example, if the controller passes `Item.limit(10)`, and the first 10 items are not available, you'll have 0 items to show. The controller should pass `Item.available.limit(10)`

File: app/controllers/purchases_controller.rb [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+  def find_item(id)
+    Item.find(id)
+  end
+
+  def item_id
+    params[:id]
+  end
+
+  def send_email(purchase_details)
+    ItemSoldMailer.notify_seller(purchase_details).deliver_now
```
Comment (kevbuchanan) [inline app/controllers/purchases_controller.rb]: This is something that can fail for network issues, and doesn't need to be done right away. For that reason you should offload this to a Sidekiq or Resque job so that it can be retried and so that it doesn't hold up the execution of this request. You can do that through ActionMailer's integration with Sidekiq or Resque through ActiveJob, or you can have a custom Sidekiq or Resque job to do this. Either way, you'll need to get redis set up as a dependency of your app, and  choose whether to use Sidekiq or Resque to do this. I don't care which one you use (I prefer Sidekiq), but you should look at both and understand the key difference(s).

File: app/views/items/index.html.erb [hunk 9]
```diff
@@ -1,13 +1,21 @@
...
+</div>
 
+<div class="container">
+  <% @presenter.available_items.each do |item| %>
+    <div class="col-xs-12 col-md-4 col-lg-3">
+      <div class="row, item-container">
+        <div class="col-md-12, item-details-container">
+          <div  class="items-index-image-container">  <%= image_tag(item.image, size: "100") %> </div>
+          <div class="title-price"> <strong> <%= item.name.capitalize %> : $<%= item.price %> </strong></div>
+          <p> <a class="btn btn-primary btn-sm btn-buy-now" href="items/<%= item.id %>/purchases/new" role="button">Buy Now!</a></p>
```
Comment (kevbuchanan) [inline app/views/items/index.html.erb]: I think [link_to](http://api.rubyonrails.org/classes/ActionView/Helpers/UrlHelper.html#method-i-link_to) is the Rails way to create this button...?

File: spec/models/purchase_spec.rb [hunk 10]
```diff
@@ -0,0 +1,41 @@
+require 'rails_helper'
+
+describe Purchase do
+
+  before(:each) do
```
Comment (kevbuchanan) [inline spec/models/purchase_spec.rb]: Please refactor this into a `let`. The before isn't necessary, and it's also not necessary to query the first Purchase when you can just do:

``` ruby
@purchase = Purchase.create({email: "" })
```

File: spec/models/purchase_spec.rb [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+    it "doesn't save without an item id" do
+      purchase = create_invalid_purchase(email: "us@example.com")
+
+      expect(purchase.errors).to include(:item_id)
+    end
+  end
+
+  private
+
+  def create_invalid_purchase(purchase_details)
```
Comment (kevbuchanan) [inline spec/models/purchase_spec.rb]: This method is not titled appropriately, there is nothing about it that creates an invalid purchase. The validity of a purchase is defined by the attributes it's created with. I'd prefer to just see `Purchase.create(your_purchase_data)` in each spec. There is nothing about pulling that into a method that increases readability.

Comment (kevbuchanan) [inline spec/models/purchase_spec.rb]: It would only warrant pulling into a method if the method did something other than just delegating to `Purchase.create`, like providing default params:

``` ruby
def create_purchase(attrs)
  Purchase.create(default_attrs.merge(attrs))
end
```

File: app/controllers/items_controller.rb [hunk 12]
```diff
@@ -24,4 +24,8 @@ def create
   def item_params
     params.require(:item).permit(:name, :description, :price, :email, :image)
   end
+
+  def available_items
+    Item.all.where(available: true)
```
Comment (kevbuchanan) [inline app/controllers/items_controller.rb]: The `all` isn't necessary here.

Comment (kevbuchanan) [inline app/controllers/items_controller.rb]: Also, please make this a scope on Item so you can to `Item.available` http://guides.rubyonrails.org/active_record_querying.html#scopes

Comment (kevbuchanan) [inline app/controllers/items_controller.rb]: That way you hide the available attribute behind the scope.

File: app/controllers/purchases_controller.rb [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+
+  def create
+    purchase = new_purchase(email: buyer_email, item_id: item_id)
+    item = find_item(item_id)
+    @presenter = create_presenter(purchase, item)
+
+    if purchase.save
+      item.update_attributes(available: false)
+      send_email(purchase, item)
+      render :show
```
Comment (kevbuchanan) [inline app/controllers/purchases_controller.rb]: This should redirect to "purchases#show", not render show

File: app/controllers/purchases_controller.rb [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+    else
+      redirect_to "/items/#{item_id}/purchases/new"
```
Comment (kevbuchanan) [inline app/controllers/purchases_controller.rb]: This should render "new", because you probably will want to show the user what went wrong once you fix your race condition that you have here.

File: app/jobs/ebay_mailer_job.rb [hunk 15]
```diff
@@ -0,0 +1,10 @@
+require 'purchase/presenters/new_purchase_presenter'
+
+class EbayMailerJob < ActiveJob::Base
```
Comment (alexcodreanu86) [inline app/jobs/ebay_mailer_job.rb]: Where is this used?

File: config/routes.rb [hunk 16]
```diff
@@ -1,5 +1,7 @@
...
+ post "/items/:id/purchases"      =>  "purchases#create"
```
Comment (alexcodreanu86) [inline config/routes.rb]: This makes me think that a purchase belongs to an item, or that an item has many purchases.

File: db/migrate/20160519102500_add_index_to_purchases.rb [hunk 17]
```diff
@@ -0,0 +1,6 @@
+class AddIndexToPurchases < ActiveRecord::Migration
+  def change
+   add_index :purchases, [:email, :item_id], unique: true
+  end
```
Comment (alexcodreanu86) [inline db/migrate/20160519102500_add_index_to_purchases.rb]: The spacing seems a bit off here

File: spec/controllers/purchases_controller_spec.rb [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+      get :new, :id => item.id
+
+      expect(response).to render_template(:new)
+    end
+  end
+
+  context "purchase successful" do
+
+    it "response successful" do
+      post :create, :id => item.id, :purchase => {email: "someone@example.com", item_id: item.id}
```
Comment (alexcodreanu86) [inline spec/controllers/purchases_controller_spec.rb]: Could `:id` and `:item_id` ever be different?

File: spec/mailers/item_sold_mailer_spec.rb [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+    let(:purchase) { double 'purchase', email: 'sternmother@email.com', item_id: 1 }
+    let(:item) { double 'item', name: 'phone', description: 'it rings', price: 54, email: 'seller@me.com' }
+    let(:mail) { described_class.notify_seller(purchase, item).deliver_now }
+
+    it 'it has a subject' do
+      expect(mail.subject).to eq('Item Sold')
+    end
+
+    it 'it is sent to the item\'s seller email' do
+
```
Comment (alexcodreanu86) [inline spec/mailers/item_sold_mailer_spec.rb:16]: How come you have a white line here and not in the spec above?

File: spec/mailers/item_sold_mailer_spec.rb [hunk 20]
```diff
@@ -0,0 +1,36 @@
...
+      expect(mail.to).to eq(["seller@me.com"])
```
Comment (alexcodreanu86) [inline spec/mailers/item_sold_mailer_spec.rb]: Double quotes, single quotes... too many options.

File: spec/models/purchase_spec.rb [hunk 21]
```diff
@@ -0,0 +1,32 @@
+require 'rails_helper'
+
+describe Purchase do
+
+  let (:purchase) { Purchase.create(email: "me@example.com", item_id: 2) }
```
Comment (alexcodreanu86) [inline spec/models/purchase_spec.rb]: This item is only used in one of your contexts. Would it make sense to move it in there?

File: app/jobs/send_purchase_mail.rb [hunk 22]
```diff
@@ -0,0 +1,10 @@
+class SendPurchaseMail
```
Comment (kevbuchanan) [inline app/jobs/send_purchase_mail.rb:1]: Looks like you're missing tests for this.

File: app/controllers/purchases_controller.rb [hunk 23]
```diff
@@ -0,0 +1,42 @@
...
+    @presenter = NewPurchasePresenter.new(purchase, item)
+  end
+
+  def create
+    purchase = Purchase.new(email: buyer_email, item_id: params[:id])
+    Item.transaction do
+      item = Item.lock.find(params[:id])
+      @presenter = NewPurchasePresenter.new(purchase, item)
+      if purchase.save
+        item.update_attributes(available: false)
```
Comment (kevbuchanan) [inline app/controllers/purchases_controller.rb:18]: What if someone else has already purchased the item between when my page was rendered and when I submit the request to purchase? You've introduced a lock, but aren't actually using it to validate that the item is still available.

You should also provide a nicer user experience if you find that the item has already been purchased.

File: config/routes.rb [hunk 24]
```diff
@@ -1,5 +1,8 @@
 Rails.application.routes.draw do
- root 'items#index'
- get  "/items/new" =>  "items#new"
- post "/items"     =>  "items#create"
+  root 'items#index'
+  get  "/items/new"                         =>  "items#new"
+  post "/items"                             =>  "items#create"
+  get  "/items/:id/purchases/new"           =>  "purchases#new", as: 'new_purchase'
```
Comment (kevbuchanan) [inline config/routes.rb]: Rename the :id param here to :item_id to be consistent across the purchases controller.

File: app/jobs/send_purchase_mail.rb (hunk 22 above)
Comment (kevbuchanan) [inline app/jobs/send_purchase_mail.rb:1]: Still missing tests for this.

**Comment by kevbuchanan** [comment]
### To Do
- [x] Add tests for your Sidekiq job's #perform method.
- [x] Fix the :item_id param in the PurchasesController

Feel free to merge on your own once those are addressed.
